pkg/adaptor/cloud/libvirt/libvirt.go

```diff
@@ -116,55 +128,196 @@ func uploadIso(isoFile string, isoVolName string, libvirtClient *libvirtClient)
 
 }
 
-func CreateDomain(ctx context.Context, libvirtClient *libvirtClient, v *vmConfig) (result *createDomainOutput, err error) {
+func getGuestForArchType(caps *libvirtxml.Caps, arch string, virttype string) (*libvirtxml.CapsGuest, error) {
+	for _, guest := range caps.Guests {
+		if guest.Arch.Name == arch && guest.OSType == virttype {
+			return &guest, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find any guests for architecture type %s/%s", virttype, arch)
+}
 
-	v.cpu = uint(2)
-	v.mem = uint(8)
-	v.rootDiskSize = uint64(10)
+func getHostCapabilities(conn *libvirt.Connect) (*libvirtxml.Caps, error) {
+	capsXML, err := conn.GetCapabilities()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get capabilities, cause: %w", err)
+	}
 
-	exists, err := checkDomainExistsByName(v.name, libvirtClient)
+	caps := &libvirtxml.Caps{}
+	err = xml.Unmarshal([]byte(capsXML), caps)
 	if err != nil {
-		return nil, fmt.Errorf("Error in checking instance: %s", err)
+		return nil, fmt.Errorf("unable to unmarshal capabilities, cause: %w", err)
 	}
-	if exists {
-		logger.Printf("Instance already exists ")
-		return &createDomainOutput{
-			instance: v,
-		}, nil
+
+	return caps, nil
+}
+
+func lookupMachine(machines []libvirtxml.CapsGuestMachine, targetmachine string) string {
+	for _, machine := range machines {
+		if machine.Name == targetmachine {
+			if machine.Canonical != "" {
+				return machine.Canonical
+			}
+			return machine.Name
+		}
 	}
+	return ""
+}
 
-	rootVolName := v.name + "-root.qcow2"
-	err = createVolume(rootVolName, v.rootDiskSize, libvirtClient.volName, libvirtClient)
+// getCanonicalMachineName returns the default (canonical) name of the guest machine based on capabilities
+// this is equivalent to doing a `virsh capabilities` and then looking at the `machine` configuration, e.g. `<machine canonical='s390-ccw-virtio-rhel9.0.0' maxCpus='248'>s390-ccw-virtio</machine>`
+func getCanonicalMachineName(caps *libvirtxml.Caps, arch string, virttype string, targetmachine string) (string, error) {
```

Comment:
I am just curious why you need the canonical machine name since that `s390-ccw-virtio` points to `s390-ccw-virtio-rhel9.0.0`. I understand when the user use the canonical name to pick a old versiion, e.g., `s390-ccw-virtio-rhel8.4.0`

Reply:
when setting up the machine name I think it's better to make the configuration of the domain XML as explicit as possible, instead of relying on defaults. Also:
- I know from several other projects that the domain creation on s390x is working if I set the canonical machine name the way you find in the code. It is possible that it is also working if I leave it out, but I suggest to have a working version first and if we find that we should simplify, we can do this on top of a working target more easily
- if the code is unable to locate a `canonicalmachine` then it will fail fast, so it would detect a configuration mismatch.